Simplifying TexTemplates and Creating Template Library

cobordism opened the pull request:
Starting work towards:
https://github.com/ManimCommunity/manim/issues/467
but also towards simplifying TexTemplates.

TexTemplate() and related classes
-----------------------------------------
Creates `TexTemplate()` class that manages tex templates. 
TexTemplate() is the default template which loads all the packages as the default 3b1b template. BasicTexTemplate() meanwhile creates a tex template that only loads basic AMS packages.

A TexTemplate object has methods like `add_to_preamble(txt)` or `add_to_document(txt)` that allow you to modify the template to fit your needs.

Finally there is the TexTemplateFromFile() class that creates a TexTemplate() object from a file and this is the class used when manim is called with the --tex_template CLI argument. 


Each `TexTemplate` has a `tex_compiler` attribute and an `output_format` attribute. The default is to use the `latex` compiler to get a `.dvi` output, but the system (implemented in tex_file_writing.py) is capable of using `xelatex` to get `.xdv` output or `lualatex` to get `.pdf` output and any combination thereof.



The Template Library
----------------------------

In addition to basic functionality, this PR also introduces ready-to-use tex template objects.
They are collected under the names `TexTemplateLibrary` and `TexFontTemplates`.

The TexTemplateLibrary contains just 4 (well 3) templates at the moment.
```
TexTemplateLibrary.default # the default template
TexTemplateLibrary.threeb1b # the 3b1b template (currently the same as the default)
TexTemplateLibrary.ctex # the 3b3b template with the use_ctex flag set
TexTemplateLibrary.simple # the simple template with only AMS packages.
```

The TexFontTemplates collection meanwhile contains all the font configurations that support maths-mode as described here: http://jf.burnol.free.fr/showcase.html
It contains 62 templates ready to use.
```
TexFontTemplates.american_typewriter,  # "American Typewriter"
TexFontTemplates.antykwa,  # "Antykwa Półtawskiego (TX Fonts for Greek and math symbols)"
TexFontTemplates.apple_chancery,  # "Apple Chancery"
TexFontTemplates.auriocus_kalligraphicus,  # "Auriocus Kalligraphicus (Symbol Greek)"
TexFontTemplates.baskervald_adf_fourier,  # "Baskervald ADF with Fourier"
TexFontTemplates.baskerville_it,  # "Baskerville (Italic)"
TexFontTemplates.biolinum,  # "Biolinum"
TexFontTemplates.brushscriptx,  # "BrushScriptX-Italic (PX math and Greek)"
TexFontTemplates.chalkboard_se,  # "Chalkboard SE"
TexFontTemplates.chalkduster,  # "Chalkduster"
TexFontTemplates.comfortaa,  # "Comfortaa"
TexFontTemplates.comic_sans,  # "Comic Sans MS"
```
 and so on.... (complete list is visible in example_scenes/advanced_tex_fonts.py and in utils/tex_templates.py)

Further changes
----------------------
`Tex` mobjects and `MathTex mobjects` no longer have the `type` attribute (in CONFIG) but instead come with a `tex_environment` attribute.
Thus, what used to be type=text is now tex_environment=None and what used to be type=tex is now tex_environment="align*".
This approach makes more sense (it is easier to understand what tex_environment means than what type mean) and it allows for more flexibility as it is now easy to create Tex Mobjects for other tex environments like arrays, eqnarrays, displaymath, ... .

cobordism commented:
I think I'm ready for some feedback.

PhilippImhof reviewed (changes requested):
This all looks good to me and works fine on my machine (Mac), at least now that I have updated my TeX distribution.

You are removing a lot of old documentation. Try to use these as a base in order to document your methods.

leotrs commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,165 @@
 
-class TexTemplateFromFile:
+class TexTemplate:
+    """
+    Class representing a TeX template to be used for creating Tex() and MathTex() objects.
```
```suggestion
    """A TeX template to be used for creating Tex() and MathTex() objects."""
```

leotrs reviewed (commented):
This looks super super super good, and it looks like a lot of effort too. Thanks! 

My only comment is that the docstrings for your classes are a bit sparse. The example scenes and examples docs are great. But just by looking at the file structure, the class names, and the class doscstrings, I can't tell which one is which and what for. For example, the `manim.utils.tex` contains three classes with `Template` in their name, but there is also another module `manim.utils.tex_templates`. Why are these "template" classes *not* part of the "template" module?

I think a lot of this vagueness can be fixed by adding more descriptive docstrings to the modules and classes, as well as "See Also" sections and short "Examples" sections. Or even just add something like "This is a base class, if you are trying to change font, look into SuchAndSuchClass instead". cc @behackl

cobordism commented:
> the manim.utils.tex contains three classes with Template in their name, but there is also another module
> manim.utils.tex_templates. Why are these "template" classes not part of the "template" module?

Well the manim.utils.tex is where the functionality of Tex Templates is defined. Creating basic templates. Modifying templates. Creating templates from files.

manim.utils.tex_templates contains actual tex templates - the objects - and a whole lot of them.
I could rename it from tex_templates to Tex_template_library. Do you think that would clarify things?

I'll add some more docstrings. TBH I only learned about how to set them yesterday :)

cobordism commented:
I have docstrings for every new class and I have added more docstrings to the methods.
There are also extensive comments in the tex_templates.py file and I have added examples to the documentation that make use the the TexTemplate class and methods and also the template library.

I need some more advice. How shall I proceed?

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,203 @@
-        after all other packages. Sometimes, the order of
-        the ``\\usepackage`` directives is relevant.
+    def prepend_to_preamble(self, txt):
+        """Adds txt to the TeX template preamble just after the \documentclass
```
```suggestion
        """Adds stuff (e.g. definitions, packages) to the TeX template preamble just after the \\documentclass
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,203 @@
 
-    def append_to_preamble(self, text):
-        """Adds commands (e.g. macro definitions) at the end of the preamble.
+    def add_to_preamble(self, txt):
+        """Adds txt to the TeX template preamble just before \\begin{document}
```
```suggestion
    def append_to_preamble(self, txt):
        """Adds stuff (e.g. definitions, packages) to the TeX template preamble just before \\begin{document}
```

The other part is called `prepend`, so I guess it would be good to use `append` here.

cobordism replied:
I though about that, but then I thought adding a package to the preamble is something that may get used once in a while, while prepending to the preamble is really a very special use function with only the rarest of uses.
add to preamble and add to document are thus the generic cases and prepend is special. no?

cobordism replied:
.... I'm not strongly opposed or anything, but if you change it here, you have to change it in the example_scenes in the tex_templates and in the examples in the documentation.

PhilippImhof replied:
I am fine with both of them. I see an advantage in the sense that `append` indicates (without looking at the docs) where the code is added. This can be important in some cases, e.g. `hyperref` -- which is of course a bad example, because we probably do not use it in a manimation.

PhilippImhof replied:
> if you change it here, you have to change it in the example_scenes in the tex_templates and in the examples in the documentation.

Absolutely. I didn't want to suggest that as long as it was not clear whether the name change was deemed useful. :-)

cobordism replied:
any solution is fine by me including add_to_preamble(txt, prepend=true) if you feel so inclined :D
knock yourself out!

but tbh saying "using `add_to_preamble` is how you add an extra latex packed" works 99% of the time so...

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,203 @@
-        """
-        Function for creating LaTeX command strings with or without options.
-        Internally used to generate ``\\usepackage`` commands.
+    def add_to_document(self, txt):
+        """Adds txt to the TeX template just after \\begin{document}, e.g. ``\\boldmath``
```
```suggestion
    def prepend_to_document(self, txt):
        """Adds stuff (e.g. definitions, macros) to the TeX template just after \\begin{document}, e.g. ``\\boldmath``
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,203 @@
+    def prepend_to_preamble(self, txt):
+        self.file_not_mutable()
+
+    def add_to_preamble(self, txt):
```
```suggestion
    def append_to_preamble(self, txt):
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,203 @@
+    def add_to_preamble(self, txt):
+        self.file_not_mutable()
+
+    def add_to_document(self, txt):
```
```suggestion
    def preprend_to_document(self, txt):
```

PhilippImhof reviewed (changes requested):
Very good job. I have only a few details left (see my suggestions). Maybe someone can have a closer look at the docstrings, e.g. @PgBiel ?

cobordism commented:
Ok. I think I'm done here.

@PhilippImhof ?

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
-        """Adds a new package (or several new packages)
-        after all other packages. Sometimes, the order of
-        the ``\\usepackage`` directives is relevant.
+    def add_to_preamble(self, txt):
```
```suggestion
    def add_to_preamble(self, txt, prepend=False):
        """Adds stuff to the TeX template's preamble (e.g. definitions, packages). Text can be inserted at the beginning or at the end of the preamble.

        Parameters
        ----------
        txt : :class:`str`
            String containing the text to be added, e.g. ``\\usepackage{hyperref}``
        prepend : Optional[:class:`bool`], optional
            Whether the text should be added at the beginning of the preample, i.e. right after ``\\documentclass``. Default is to add it at the end of the preample, i.e. right before ``\\begin{document}``
        """
        if prepend:
            self.preamble = txt + "\n" + self.preamble
        else:
            self.preamble += "\n" + txt
        self.rebuild()
```

Liked your suggestion of having just one method for both cases. Sorry, review comments are a bit messy, but it seems impossible to make a suggestion when there are deleted lines in between.

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
-        """Adds a new package (or several new packages)
-        before all other packages. Sometimes, the order of
-        the ``\\usepackage`` directives is relevant.
+    def prepend_to_preamble(self, txt):
+        """Adds txt to the TeX template preamble just after the \documentclass
 
         Parameters
         ----------
```
```suggestion

```

Not needed anymore if we have `add_to_preample` with parameter for prepending/appending.

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
         ----------
-        pkg : :class:`str`
-            The package name, e.g. ``"siunitx"``
+        txt : :class:`str`
+            String containing the text to be added, e.g. ``\\usepackage{siunitx}``
         """
```
```suggestion
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
         """
-        self.common_packages.insert(0, pkg)
-        self.rebuild_cache()
+        self.preamble = txt + "\n" + self.preamble
+        self.rebuild()
 
```
```suggestion
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
-        after all other packages. Sometimes, the order of
-        the ``\\usepackage`` directives is relevant.
+    def add_to_preamble(self, txt):
+        """Adds txt to the TeX template preamble just before \\begin{document}
```
```suggestion
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
-        the ``\\usepackage`` directives is relevant.
+    def add_to_preamble(self, txt):
+        """Adds txt to the TeX template preamble just before \\begin{document}
 
         Parameters
         ----------
```
```suggestion
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
         ----------
-        pkg : :class:`str`
-            The package name, e.g. ``"siunitx"``
+        txt : :class:`str`
+            String containing the text to be added, e.g. ``\\usepackage{hyperref}``
         """
```
```suggestion
```

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,189 @@
         """
-        self.common_packages.append(pkg)
-        self.rebuild_cache()
+        self.preamble += "\n" + txt
+        self.rebuild()
 
```
```suggestion
```

PhilippImhof reviewed (changes requested):
I like your idea of having just one method for adding stuff to the preamble. Everything else already looked good to me. Suggestions got a bit messy, but once they are done, we can easily run another test.

PhilippImhof reviewed (approved):
Very good job. Thanks a lot.

PgBiel commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,179 @@
-        return self.body.replace(
-            self.text_to_replace, "{0}\n{1}\n{2}".format(begin, expression, end)
+class TexTemplate:
+    """TeX templates are used for creating Tex() and MathTex() objects."""
```
please read our guidelines on class docstrings. This is lacking the attributes and stuff

behackl replied:
We should have this discussion sooner or later, and this reminds me: I would like to get rid of typing in the docstrings (while I do think that it improves documentation, it is **hard** to get right especially for people not into Sphinx/docstrings), and slowly introduce type hints (via PEP 484, https://www.python.org/dev/peps/pep-0484/). 

Sphinx is able to generate the same type information from type hints in the code as from the Numpy-style annotations in the docstring -- but IDEs can profit more from type hints.

What do you think, @PgBiel?

PgBiel replied:
The point of typing is to have docstrings be as precise as possible. And, most of what you'll need to write typings is in our guidelines.

Thing is, we do want to add typings; however, that requires people to contribute with typings on every PR, in order to not increase our workload indefinitely. So, whether we like it or not, people will need to write them. Enforcing typings on the docstring makes our job easier later. However, people may also write typehints and have us transport them to the docstring if they want an easier time (although they may do that by themselves).

So the original idea is to have both, in order to have things more organized.

PhilippImhof replied:
> Thing is, we do want to add typings; however, that requires people to contribute with typings on every PR, in order to not increase our workload indefinitely. 

I think it just has to be clear. Because in #457 just a few days ago, the position seemed to be different. If I may quote @leotrs:

> I think the typings can be left for a future PR. No need to request that right now, when we are not requesting the same from any other PR.
>
> In fact, there's an entirely different branch that is taking care of typings.

or @behackl:

> Yes, please, lets not overthink this. We don't have in-code typings anywhere in the code base yet, now and this PR is not the time and place to start introducing them.

Also, we should ask ourselves whether this might deter people from contributing valuable ideas, because they think they need to send "perfect" code. Adding typings can also be done later, it can be added as an easy issue. Anyhow, I will gladly help in adding needed typings.

PhilippImhof replied:
BTW: The Guidelines (https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP)#numpy-format) say: 

> `# typings are optional for now`

leotrs replied:
To put context to that quote: someone had suggested we add typings on a PR that i) had been open for a while, and ii) it was suggested right at the last minute, once all other issues had been dealt with. At that point, my opinion was (and still is) that typings were not necessary to merge that PR.

Having said that, I think we should incorporate typings gradually, but we should not block PRs because they don't have typings. Usually I try to request some minimal documentation in each PR, but I think that filling in most sections of each docstring, as well as typings, can be left for the future.

I agree in general we should not turn away good/new ideas just because they lack typings.

PgBiel replied:
Thing is that, if we do plan on adding typings, then we need to start doing so now, because, otherwise, our workload will just increase indefinitely, as I said. Trust me, it's not that simple to start typing a repo with 0 types. So, by enforcing this, our job is way easier. By doing this, we're thinking of the future.

We don't have to add typehints in the code yet, but at least in the docstrings so that we know what to do later.

And regarding the guidelines: sorry if that wasn't clear; I meant that _typehints_ (i.e., in the code) are optional. Note the example that you quoted: it doesn't have any typings in its parameters, but it does have in the docstring. I'll update the guidelines to make that clear.

behackl replied:
I do know why typing is useful and I am not against having this kind of precision in our code base; far from it!

But at the same time, we have to be realistic. Adding Numpy-style docstrings, even though all of it is documented in our Wiki, is difficult to get right: the syntax is rather particular and unforgiving, and there is (at least in my IDE) no indication of whether the format I am writing is correct, leading to frustration for both contributors and reviewers.

This is why I would like to see us moving from typings in the docstrings to type hints, directly in the code (and most modern IDEs support checking your added type hints in some way, in particular whether they are syntactically correct). Adding these correctly is a lot easier (because the parser shouts at you) -- and thus also lowers the participation barrier for new (and existing) contributors.

I don't really see why you would like to have both type hints in the method signature and typings in the docstring: to be frank, I think wanting to have both is simply a disaster waiting to happen from a maintainance point of view. The rendered HTML documentation contains the same level of precision in both scenarios, and inspection shows PEP 484 type hints even better than typings added in the docstring.

Overall, I think type hints are easier to maintain than the docstring typings while, at the same time, yielding the same information. Whatever this discussion (that should probably be relayed to a new (or old?) issue) results in: I am strictly against maintaining typings in both the function signature via type hints and the docstring. I don't see how this makes sense or improves the available information in any way.

leotrs replied:
Agreed we should start somewhere. However, this PR has been open for almost a month now so I'm not sure this is the right place to start doing so. It can be very annoying for a dev to be requested to add typings only after three weeks of working on this.

I think we should have a separate discussion about what to do with typings. Honestly, I think we can do without them, just like @behackl said, but @PgBiel has pushed hard for them in the past and I respect that.

PhilippImhof replied:
Thanks @PgBiel for the clarification. This leads me to another question: How and why did we get to the typing question here? It seems to me that the docstrings -- while lacking attrs at the class level --  contain all necessary typings. That's why I thought you were talking about typehints. (And because I was not aware of the subtle difference between these two terms.)

I will try to give @cobordism a hand for the class docstring.

behackl replied:
I started this, sorry. Let us take this to a dedicated issue. :-)

eulertour replied:
After following the activity in this PR for a while I also think we need a better approach for this.

PgBiel replied:
@PhilippImhof yes, the docstrings should contain all necessary typings for now. Thing is, @behackl suggested we use typehints instead.

Here's my take on this: We can make typings on docs simpler. However, we may only do that _after_ we work on typings. Because we need at least _some way_ to deduce typings for parameters and stuff when we do it. Then, after said work is done, we may begin to enforce typehints for new PRs.

Also, note: You said that adding typehints makes the IDE parse and stuff. You can write them in the code, have the IDE suggest the changes and _then_ you add the typings to the docstring.

behackl replied:
Lets take this discussion to #52.

PhilippImhof replied:
@PgBiel 

> Here's my take on this: We can make typings on docs simpler. However, we may only do that _after_ we work on typings. 

I wouldn't want to be annoying, but are you talking about typings and typings or typings and typehints? I am just trying to use those terms correctly in the future.

> Also, note: You said that adding typehints makes the IDE parse and stuff. You can write them in the code, have the IDE suggest the changes and _then_ you add the typings to the docstring.

That wasn't me. 😄  My IDE is Emacs, it doesn't do anything like that. (It certainly could do it, but I do not have the necessary packages installed.)

PgBiel replied:
> I wouldn't want to be annoying, but are you talking about typings and typings or typings and typehints? I am just trying to use those terms correctly in the future.
Oh, I'm talking about typehints, yeah. And this part of the reply was meant for @behackl 

> That wasn't me.
Woops, sorry, I meant to ping @behackl 😅 sorry for that

PgBiel replied:
Any update on this? Documenting attrs and init params is necessary

PgBiel commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,179 @@
+        )
+        self.rebuild()
+
+    def rebuild(self):
```
no docs?

cobordism replied:
If you want to use a template, you mightcreate a TexTemplate() and you might add_to_preamble(), but you don't need to call rebuild().

We have to decide at some point: who is the documentation for?
Are we documenting for people who want to use manim or for people who want to hack manim? The two are not the same, but, by nature of this project, they are not clearly separated either.

This function is only useful for people who want to hack away at the template system, not for people who are using manim.

leotrs replied:
Ideally, documentation could be used by either users or developers. I do agree that user-docs should take precedence over dev-docs.

However, with the current state of things, *any* docs are better than none.

cobordism replied:
or rename it to `_rebuild()` and just 'document' it with a # comment ?

leotrs replied:
Yes for making it private, but even private functions deserve a docstring. Whatever you would put in a comment, put it in a docstring.

cobordism replied:
ok. I'll do that quickly. hang on.

PgBiel commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,179 @@
         -------
         :class:`str`
-            LaTeX code based on the template, containing the given expression and ready for typesetting
+            LaTeX code based on template, containing the given expression and ready for typesetting
```
why was "the" removed here? Lol
Either "the" or "a", pick one

cobordism replied:
"template" here is the argument to the function.

LaTeX code based on **template**, containing the given **expression** and ready for typesetting

cobordism replied:
and it wasn't removed per se. The whole file was written from scratch.. it's just github that insists on showing this as a series of diffs.

PhilippImhof replied:
Is there some qualifier similar to `:attr:` so we could write "based on `:param:template` or something like that?

PgBiel replied:
Ahhh I see, then just surround "template" by double-backticks (\`\`template\`\`) so that users become aware that you refer to the parameter.

cobordism replied:
actually, I think I was mistaken here.The template in question is `self`. I have reworded it.

PgBiel commented on manim/utils/tex_file_writing.py:
```diff
@@ -21,21 +21,56 @@ def tex_hash(expression):
+
+    Returns
+    -------
+    :class:`string`
```
```suggestion
    :class:`str`
```

PgBiel replied:
??? this was not resolved

PhilippImhof replied:
As I understood, @cobordism preferred to copy those suggestions over to his local documents and commit rather than to use the GitHub suggestion features. This one was resolved in commit 6720c86c

PgBiel replied:
oh, it was outdated for me

PgBiel commented on manim/utils/tex_templates.py:
```diff
@@ -0,0 +1,1022 @@
+    """
+    A collection of basic TeX template objects
+
+    Usage
```
```suggestion
    Examples
```

PgBiel commented on manim/utils/tex_templates.py:
```diff
@@ -0,0 +1,1022 @@
+
+    Usage
+    -----
+    As a value for the keyword argument tex_template of Tex() and MathTex() mobjects
```
```suggestion
    As a value for the keyword argument tex_template of Tex() and MathTex() mobjects::

```

PgBiel commented on manim/utils/tex_templates.py:
```diff
@@ -0,0 +1,1022 @@
+    """
+
+    default = TexTemplate()
+    """ An instance of the default TeX template in manim """
```
the guidelines recommend bundling the attrs in the class docstring, but this particular case is debatable

cobordism replied:
"bundling the attrs in the class docstring"
I have to admit I have no idea what these words mean.

PgBiel replied:
i.e., having one Attributes section in the class's docstring, instead of distributing the docstrings of each attribute to under them

cobordism replied:
ah yes. I see. Makes sense. Noted for future.
But with these particular classes (which just serve as a collection of attributes) perhaps that is suboptimal.

PgBiel replied:
yeah, I guess we can open an exception here then

PgBiel replied:
let's remove those spaces
```suggestion
    """An instance of the default TeX template in manim"""
```

PgBiel commented on manim/utils/tex_templates.py:
```diff
@@ -0,0 +1,1022 @@
+    These templates are specifically designed to allow you to typeset formulae and mathematics using
+    different fonts. They are based on the mathastext LaTeX package.
+
+    Usage
```
also goes after Notes i believe
```suggestion
    Examples
```

cobordism replied:
I meant usage as in "this is how it is used".
Namely (next line) "As a value for the keyword argument tex_template of Tex() and MathTex() mobjects"

And then I give a one line example just to show precisely how I meant that. (i.e. the syntax)

Calling this section 'examples' doesn't feel quite right to me.

PgBiel replied:
oh, this is just because there is no Usage section in the numpy format

cobordism replied:
I'm sorry. I didn't know there even was such a thing as "numpy format".

I learned the other day that I should change my `#` comment documentation into `"""` comment documentation and that these comments should contain restructured text (rst).

PgBiel replied:
haha dont worry, it's alright! The guidelines should be helpful here then. Make sure to take a look!

PgBiel replied:
this still wasnt resolved; i guess we should just add this to the long description then?

PhilippImhof commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,179 @@
+        )
+        self.rebuild()
+
+    def rebuild(self):
```
```suggestion
    def rebuild(self):
        """Rebuilds the entire TeX template text from ``\\documentclass`` to ``\\end{document}`` according to all settings and choices."""
```

cobordism commented:
Ready for final(*) review.


(*) hopefully

PgBiel commented on manim/utils/tex.py:
```diff
@@ -1,298 +1,179 @@
-            The string containing the (math) expression to be typeset,
-            e.g. ``"$\\sqrt{2}$"``
+class TexTemplateFromFile(TexTemplate):
+    """A TexTemplate object created from a template file (default: tex_template.tex)"""
```
attributes and init params

PhilippImhof replied:
These are all resolved, again by a new commit (https://github.com/ManimCommunity/manim/pull/483/commits/79d0d7c9da88b831920ed42a5e97962ab5615796)

PgBiel commented on manim/utils/tex_file_writing.py:
```diff
@@ -21,21 +21,56 @@ def tex_hash(expression):
+        String containing the TeX expression to be rendered, e.g. ``\\sqrt{2}`` or ``foo``
+    environment : Optional[:class:`str`], optional
+        The string containing the environment in which the expression should be typeset, e.g. ``align*``
+    tex_template : Optional[:class:`TexTemplate`], optional
```
TexTemplate is in another file, so we must tell sphinx to search for it using `.`; `~` to shorten to `TexTemplate` instead of the full path
```suggestion
    tex_template : Optional[:class:`~.TexTemplate`], optional
```

PgBiel commented on manim/utils/tex_file_writing.py:
```diff
@@ -21,21 +21,56 @@ def tex_hash(expression):
+        String containing the TeX expression to be rendered, e.g. ``\\sqrt{2}`` or ``foo``
+    environment : Optional[:class:`str`], optional
+        The string containing the environment in which the expression should be typeset, e.g. ``align*``
+    tex_template : Optional[:class:`TexTemplate`], optional
```
```suggestion
    tex_template : Optional[:class:`~.TexTemplate`], optional
```

PgBiel commented on manim/utils/tex_file_writing.py:
```diff
@@ -75,53 +128,71 @@ def tex_compilation_command(compiler, tex_file, tex_dir):
+    Returns
+    -------
+    :class:`str`
+        Path to generated SVG file."""
```
just for style etc 
```suggestion
        Path to generated SVG file.
        """
```

PgBiel commented on manim/utils/tex_templates.py:
```diff
@@ -0,0 +1,1022 @@
+            r"\DisableLigatures{encoding = *, family = * }", r"\usepackage[UTF8]{ctex}"
+        ),
+    )
+    """ An instance of the TeX template used by 3b1b when using the use_ctex flag """
```
```suggestion
    """An instance of the TeX template used by 3b1b when using the use_ctex flag"""
```

PgBiel commented on manim/utils/tex_templates.py:
```diff
@@ -0,0 +1,1022 @@
+    """ An instance of the TeX template used by 3b1b when using the use_ctex flag """
+
+    simple = _new_ams_template()
+    """ An instance of a simple TeX template with only basic AMS packages loaded """
```
```suggestion
    """An instance of a simple TeX template with only basic AMS packages loaded"""
```

cobordism commented:
> this still wasnt resolved; i guess we should just add this to the long description then?

what do you mean? Is this still not correct?
https://github.com/cobordism/manim/blob/refactor-tex-templates/manim/utils/tex_templates.py#L904

cobordism commented:
This time I clicked to accept the suggestions in github web interface instead of submitting a commit with the changes myself
... and now black fails.
I'm tired of this. Please either merge or reject... do we really need all this back and forth about a newline character in a comment string?
https://github.com/ManimCommunity/manim/pull/483/commits/73373595fa2e8987ce76842117592ddd7554b784

PhilippImhof commented:
> This time I clicked to accept the suggestions in github web interface instead of submitting a commit with the changes myself
> ... and now black fails.

Probably the wrong indentation for the closing triple quote on L187

leotrs commented:
Don't touch anything for a sec, I'm about to merge this

leotrs commented on manim/utils/tex_file_writing.py:
```diff
@@ -75,53 +128,70 @@ def tex_compilation_command(compiler, tex_file, tex_dir):
+    Returns
+    -------
+    :class:`str`
+        Path to generated SVG file.
+        """
```
```suggestion
        Path to generated SVG file.
    """
```

leotrs commented:
I count two approvals now. Any documentation tweaks can be made later on.

PgBiel commented:
> > this still wasnt resolved; i guess we should just add this to the long description then?
> 
> what do you mean? Is this still not correct?
> https://github.com/cobordism/manim/blob/refactor-tex-templates/manim/utils/tex_templates.py#L904

huh, weird, it still appeared as "Usage" for me; now it is correct. Sorry!

PgBiel commented:
> Any documentation tweaks can be made later on.

We have to be careful with this approach, as we can't just pile up endless work for ourselves. We need to make sure each PR has appropriate documentation.

This one seems to be reasonably good already, however, given the multiple reviews.

leotrs commented:
I agree with you. And I also agree with @cobordism in that going back and forth about single blank lines seems a bit excessive. We need better ways of doing it. I keep wishing for a docstring autoformatter, like black but for docstrings.

PhilippImhof commented:
What about this:

https://stackoverflow.com/questions/42123798/is-there-any-way-to-lint-a-python-file-to-check-for-numpy-documentation-style-ad

leotrs commented:
That looks good. If anyone could try that out locally and see how that works that'd be great. cc @ManimCommunity/docs

kolibril13 commented:
```py
class InCodeTexTemplate(Scene):
    """This example scene demonstrates how to modify the tex template
    for a particular scene from the code for the scene itself.
    """

    def construct(self):
        # Create a new template
        myTemplate = TexTemplate()

        # Add packages to the template
        myTemplate.add_to_preamble(r"\usepackage{esvect}")

        # Set the compiler and output format (default: latex and .dvi)
        # possible tex compilers: "latex", "pdflatex", "xelatex", "lualatex", "luatex"
        # possible output formats: ".dvi",  ".pdf", and ".xdv"
        myTemplate.tex_compiler = "pdflatex"
        myTemplate.output_format = ".pdf"

        # To use this template in a Tex() or MathTex() object
        # use the keyword argument tex_template
        text = MathTex(r"\vv{vb}", tex_template=myTemplate)
        self.play(Write(text))
        self.wait(1)
```
this script is working on my machine! :)

and this as well!
```py
class TexTemplateFromCLI(Scene):
    """This scene uses a custom TexTemplate file.
    The path of the TexTemplate _must_ be passed with the command line
    argument `--tex_template <path to template>`.
    For this scene, you can use the custom_template.tex file next to it.
    This scene will fail to render if a tex_template.tex that doesn't
    import esvect is passed, and will throw a LaTeX error in that case.
    """

    def construct(self):
        text = MathTex(r"\vv{vb}")
        self.play(Write(text))
        self.wait(1)



import os ; import sys
from pathlib import Path
if __name__ == "__main__":
    project_path = Path(sys.path[1]).parent
    script_name = f"{Path(__file__).resolve()}"
    os.system(f"manim  -l --custom_folders --tex_template /home/projects/manim-community/example_scenes/custom_template.tex --disable_caching -s -p -c 'BLACK' --config_file '{project_path}/manim_settings.cfg' " + script_name)
```
